### PR TITLE
Fix dependencies, so headers are built first

### PR DESCRIPTION
**Pull request by thomasgoirand**
There's a race condition where on slow hosts, headers aren't generated before starting the build process. The way to fix this in a Debian package is to do:

override_dh_auto_build:
        dh_auto_build --buildsystem=cmake -- legacy-option-headers
        dh_auto_build --buildsystem=cmake

but this is a hack, and best is to get these dependencies fixed in the build process directly. This is what this patch is attempting to do.

Co-Author: Xinhui Yang <cyan@cyano.uk>

### CODE REVIEW

**Comment by adamemerson** [comment]
jenkins test windows

File: src/common/options/CMakeLists.txt [hunk 1]
```diff
@@ -104,8 +103,9 @@ add_options(rgw)
 
 add_library(common-options-objs OBJECT
   ${common_options_srcs})
-add_custom_target(legacy-option-headers
-  DEPENDS ${legacy_options_headers})
+
+# Make common-options-objs as a pre-build requirement.
+add_dependencies(prebuild-targets common-options-objs)
```
Comment (tchaikov) [inline src/common/options/CMakeLists.txt:108]: @thomasgoirand hi Thomas, what these targets depend on are the option headers. and we do have a target for building them, why you prefer adding a new target and let it depend on `common-option-objs` instead? can we just s/prebiuld-targets/legacy-option-headers/ ? or you've tried it?

**Comment by tchaikov** [comment]
could you prefix the title of your commit message with the subcomponent your are changing ? see https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst#3-describe-your-changes ? in this case, it would be 'cmake:'. so the title could be `cmake: let all targets depend on prebuild-targets`

File: src/common/options/CMakeLists.txt [hunk 2]
```diff
@@ -1,5 +1,7 @@
+# A custom target using comon_options_srcs will generate YAML files and its
+# corresponding source files and header files. Remove the headers target to
+# prevent headers from generating twice.
```
Comment (tchaikov) [inline src/common/options/CMakeLists.txt:3]: i don't get this. if headers are already generated and updated, why they are generated twice?

**Comment by tchaikov** [comment]
@thomasgoirand hi Thomas, just created #57011 to prepare for the proposed fix. i agree with your approach in general, but wanted to ensure that the dependency is accurate -- the users of the options depend on the headers not on `common-options-objs`.  probably we can just link against `legacy-option-headers`? if you are with me, maybe you could cherry-pick my change and update yours to _link_ against `legacy-option-headers` instead?

**Comment by tchaikov** [comment]
@thomasgoirand ping?

**Comment by tchaikov** [comment]
@thomasgoirand hi Thomas, thanks! i will take over this patch by cherry-pick it into my branch and adapting it. will update you once i get it done. as i only work on it on weekends. so latency is expected.

**Comment by tchaikov** [comment]
closing in favor of #57011

**Pull request closed without merging**